typing/typemod.ml

```diff
@@ -557,33 +560,36 @@ let merge_constraint initial_env loc sg lid constr =
           Typedecl.transl_with_constraint id ~fixed_row_path:(Pident id_row)
             ~sig_env ~sig_decl:decl ~outer_env:initial_env sdecl in
         let newdecl = tdecl.typ_type in
-        check_type_decl sig_env sdecl.ptype_loc id row_id newdecl decl rs rem;
+        let row_id, ghosts = split_row_id s ghosts in
+        check_type_decl sig_env sdecl.ptype_loc id row_id newdecl decl
+          rec_group;
         let decl_row = {decl_row with type_params = newdecl.type_params} in
         let rs' = if rs = Trec_first then Trec_not else rs in
-        (Pident id, lid, Twith_type tdecl),
-        Sig_type(id_row, decl_row, rs', priv)
-        :: Sig_type(id, newdecl, rs, priv)
-        :: rem
-    | (Sig_type(id, sig_decl, rs, priv) :: rem , [s],
-       (With_type sdecl | With_typesubst sdecl as constr))
+        return ~ghosts
+          ~replace_by:[ Sig_type(id, newdecl, rs, priv);
+                        Sig_type(id_row, decl_row, rs', priv)
```

Comment:
If I understand correctly, here the second `Sig_type` is morally a ghost type declaration, but we don't say it explicitly because we are expanding into a non-grouped list anyway. Is that right? Would the code look nicer if we tried to reconstruct a valid group first, and then flatten it within the `replace` function?

Reply:
In fact, we need to transfer to the ghosts group to fix a bug in trunk. Currently
```ocaml
module type s = sig
  type t
  and u = private < .. >
  and v
end with type u = private < .. >
```
is printed
```ocaml
module type s = sig type t and u#row and u = private < .. > and v end
```
because the `with` constraint breaks the assumption that all ghost row components are prefixed to the group. 
I have an incoming patch that does this.